lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -257,40 +272,73 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
          "endBB is required to post-dominate startBB");
   SmallVector<std::unique_ptr<SESERegionTree>, 4> results;
 
-  // Iteratively work our way up the post-dominator tree (moving startBB until
+  // Iteratively work our way up the post-dominator tree (moving currentBB until
   // we reach the endBB), producing a sequence of loops, diamonds, and single
   // block nodes as we go.
-  while (startBB != endBB) {
+  SILBasicBlock *currentBB = startBB;
+  while (currentBB != endBB) {
+    // If currentBB is not dominated by startBB, we create a function region.
+    if (!DI.dominates(startBB, currentBB)) {
+      auto iter = functionRegions.find(currentBB);
```

Comment:
this function is getting quite long and complex -- would it be useful to extract the if body into a function?